gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/src/AuthorizationService.js

```diff
@@ -662,42 +712,125 @@ var fluid = fluid || require("infusion");
+        var accessToken = codeGenerator.generateAccessToken();
+
+        var addResourceOwnerTokenPromise = dataStore.addResourceOwnerToken({
+            clientId: clientId,
+            gpiiToken: gpiiToken,
+            accessToken: accessToken,
+            expiresIn: expiresIn
         });
+
+        var mapper = function () {
+            // The created access token is resolved for promiseTogo eventually
+            return {
+                accessToken: accessToken,
+                expiresIn: expiresIn
+            };
+        };
+        var accessTokenPromise = fluid.promise.map(addResourceOwnerTokenPromise, mapper);
+        fluid.promise.follow(accessTokenPromise, promiseTogo);
+    };
+
+    /**
+     * Grant a resource owner password credential access token. The gpii token will be verified before the access token is returned.
+     * @param dataStore {Component} An instance of gpii.oauth2.dbDataStore
+     * @param codeGenerator {Component} An instance of gpii.oauth2.codeGenerator
+     * @param clientId {String} A client id
+     * @param gpiiToken {String} A GPII token
+     * @return {Promise} A promise object whose resolved value is the access token. An error will be returned if the gpii token is not found.
+     */
+    gpii.oauth2.authorizationService.grantResourceOwnerAccessToken = function (dataStore, codeGenerator, gpiiToken, clientId) {
+        var promiseTogo = fluid.promise();
+
+        if (!gpiiToken || !clientId) {
+            var error = gpii.oauth2.composeError(gpii.oauth2.errors.missingInput, {fieldName: "GPII token or client ID"});
+            promiseTogo.reject(error);
+        } else {
+            var gpiiTokenPromise = dataStore.findGpiiToken(gpiiToken);
+            var clientPromise = dataStore.findClientById(clientId);
+            var resourceOwnerTokenPromise = gpii.oauth2.authorizationService.findValidResourceOwnerToken(dataStore, gpiiToken, clientId);
+
+            // TODO: Update the usage of fluid.promise.sequence() once https://issues.fluidproject.org/browse/FLUID-5938 is resolved.
+            var sources = [gpiiTokenPromise, clientPromise, resourceOwnerTokenPromise];
+            var promisesSequence = fluid.promise.sequence(sources);
+
+            promisesSequence.then(function (responses) {
+                var gpiiTokenRec = responses[0];
+                var clientRec = responses[1];
+                var resourceOwnerTokenRec = responses[2];
+
+                var error;
+
+                if (!gpiiTokenRec) {
+                    error = gpii.oauth2.composeError(gpii.oauth2.errors.invalidGpiiToken);
+                    promiseTogo.reject(error);
+                } else if (clientRec.oauth2ClientType !== gpii.oauth2.oauth2ClientTypes.nativeApp) {
+                    error = gpii.oauth2.composeError(gpii.oauth2.errors.unauthorizedClient);
+                    promiseTogo.reject(error);
```

Comment:
Are the error messages for invalidGpiiToken and unauthorizedClient sent in the HTTP response? If so, I wonder if differentiating between these 2 unauthorized cases is the right thing to do. It certainly helps in debugging and diagnosing issues and it's probably something we should log. But, the more information we provide in a response for the reason that a request was rejected, the more it helps attackers. I've seen in some systems an id being generated for a failure -- we would make a detailed log entry indexed by the id and then the id is sent in the HTTP response. Enabling the details to be looked up but they are not public.